<p style="text-align: right; font-size: small;">
Learning materials by Tamsin Huggins
</p>

# AirBnB Challenge


Import the datafrom airbnb.json, which is in the Data folder.

In [1]:
import json
with open('../data/airbnb.json', 'r',encoding='utf-8') as f:
    data = json.load(f)


### 1) Retrieve the first AirBnB listing in the data.
This challenge is done for you to get you started.

In [2]:
data[0]

{'_id': {'$oid': '65904980a2761bfa415c2106'},
 'name': 'Home in Dublin · ★4.76 · 1 bedroom · 2 beds · 1 bath',
 'description': 'Beautiful entire house with superfast broadband and smart TV, 3 potential quiet workspaces, 10 minutes from Grafton Street, in the coolest area: Portobello, close to all the best cultural must-sees, eateries and bars, right by the leafy canal with its elegant swans. Ideal for a couple, or a single person.<br /><br /><b>The space</b><br />Gorgeous, newly revamped central Dublin city entire house. Welcome to Lennox Place, in the heart of sought-after Portobello, a chic and stylish two bedroom mid-terraced city residence, which has recently undergone a tasteful refurbishment. Beyond the classical red brick exterior now lies a bijou pied-a-terre where style, flair, light and well-appointed accommodation. Adjacent to the canal, on a Victorian tree-lined street, the house gets both morning and evening sun and is in a dream location within ten minutes walking distanc

### 2) Retrieve the description of the first AirBnB listing in the data.

In [3]:
data[0]['description']

'Beautiful entire house with superfast broadband and smart TV, 3 potential quiet workspaces, 10 minutes from Grafton Street, in the coolest area: Portobello, close to all the best cultural must-sees, eateries and bars, right by the leafy canal with its elegant swans. Ideal for a couple, or a single person.<br /><br /><b>The space</b><br />Gorgeous, newly revamped central Dublin city entire house. Welcome to Lennox Place, in the heart of sought-after Portobello, a chic and stylish two bedroom mid-terraced city residence, which has recently undergone a tasteful refurbishment. Beyond the classical red brick exterior now lies a bijou pied-a-terre where style, flair, light and well-appointed accommodation. Adjacent to the canal, on a Victorian tree-lined street, the house gets both morning and evening sun and is in a dream location within ten minutes walking distance of the city centre and so full of character - probably the most desirable, safe and interesting place to stay in the entire c

### 3) Retrieve the name of the second AirBnB listing in the data.

In [4]:
data[1]['name']

'Bungalow in Churchtown · ★4.80 · 1 bedroom · 2 beds · 1.5 shared baths'

### 4) Retrieve the reviews of the second AirBnB listing in the data.
How would you describte the structure of the result? 

In [5]:
data[1]['reviews']

{'overall': 4.8,
 'accuracy': 4.83,
 'cleanliness': 4.74,
 'checkin': 4.93,
 'communication': 4.93,
 'location': 4.66,
 'value': 4.81}

### 5) Retrieve just the location review of the second AirBnB listing in the data.

In [6]:
data[1]['reviews']['location']

4.66

### 6) What data fields (keys) are documented for each AirBnB listings? Use an dictionary method to retrieve this information.

In [7]:
data[0].keys()

dict_keys(['_id', 'name', 'description', 'host_name', 'host_is_superhost', 'neighbourhood', 'latitude', 'longitude', 'property_type', 'room_type', 'bedrooms', 'beds', 'amenities', 'price', 'minimum_nights', 'maximum_nights', 'has_availability', 'number_of_reviews', 'reviews'])

### 7) What are the possible categories on which users can review their stay? Use a dictionary method to retrieve this information.

In [8]:
data[0]['reviews'].keys()


dict_keys(['overall', 'accuracy', 'cleanliness', 'checkin', 'communication', 'location', 'value'])

### 8) Retrieve the first 3 AirBnB listings in the data.

In [9]:
data[:3]

[{'_id': {'$oid': '65904980a2761bfa415c2106'},
  'name': 'Home in Dublin · ★4.76 · 1 bedroom · 2 beds · 1 bath',
  'description': 'Beautiful entire house with superfast broadband and smart TV, 3 potential quiet workspaces, 10 minutes from Grafton Street, in the coolest area: Portobello, close to all the best cultural must-sees, eateries and bars, right by the leafy canal with its elegant swans. Ideal for a couple, or a single person.<br /><br /><b>The space</b><br />Gorgeous, newly revamped central Dublin city entire house. Welcome to Lennox Place, in the heart of sought-after Portobello, a chic and stylish two bedroom mid-terraced city residence, which has recently undergone a tasteful refurbishment. Beyond the classical red brick exterior now lies a bijou pied-a-terre where style, flair, light and well-appointed accommodation. Adjacent to the canal, on a Victorian tree-lined street, the house gets both morning and evening sun and is in a dream location within ten minutes walking dist

### 9) Retrieve the last 2 AirBnB listings in the data.

In [10]:
data[-2:]

[{'_id': {'$oid': '65904982a2761bfa415c42ec'},
  'name': 'Home in Dublin 9 · ★New · 1 bedroom · 1 bed · 1 bath',
  'description': 'Our house in the middle between airport (10-12 mins)and city center(12-15mins)<br /><br /><b>The space</b><br />Our house is semi detached house,beside a few take away pizzas, fish&chip,pharmacy,late night convenience shop,hair dresser,2 mins to Spar shop as well..we share kitchen & bathroom ,(same floor with your room )with 2 or 4 other it’s depending on booking..<br />Bus number 1 service  from city will drop u just right In front of our house..',
  'host_name': 'Jolin',
  'host_is_superhost': False,
  'neighbourhood': 'Dublin City',
  'latitude': 53.3908765,
  'longitude': -6.257394499999999,
  'property_type': 'Private room in home',
  'room_type': 'Private room',
  'bedrooms': 1,
  'beds': 1,
  'amenities': ['Backyard',
   'Dishes and silverware',
   'Free parking on premises',
   'High chair',
   'Hair dryer',
   'Washer',
   'Indoor fireplace',
   'B

### 10) Retrieve  the **names** of the first 5 AirBnB listings in the data.

You must deliver a data structure that holds **just** the name fields, which look like 'Home in Dublin · ★4.76 · 1 bedroom · 2 beds · 1 bath'

In [11]:
#comprehension solution. For loop solution also valid.
print([x['name'] for x in data[:5]])

['Home in Dublin · ★4.76 · 1 bedroom · 2 beds · 1 bath', 'Bungalow in Churchtown · ★4.80 · 1 bedroom · 2 beds · 1.5 shared baths', 'Bungalow in Churchtown · ★4.81 · 1 bedroom · 1 bed · 1.5 shared baths', 'Home in Dublin · ★4.75 · 1 bedroom · 1 bed · 2 shared baths', 'Townhouse in Dublin · ★4.85 · 2 bedrooms · 4 beds · 1.5 shared baths']


### 11) Of the first 20 listings, return the name of any that offer more than 3 beds


In [12]:
[x['name'] for x in data[:20] if x['beds'] > 3]

['Townhouse in Dublin · ★4.85 · 2 bedrooms · 4 beds · 1.5 shared baths',
 'Home in Dublin · ★4.93 · 4 bedrooms · 5 beds · 2.5 baths']

### 12) Of the first TEN listings, retrieve the name of any that have an overall review score of above or equal to 4.88.

In [13]:
[x['name'] for x in data[:10] if x['reviews']['overall'] >= 4.88]

['Home in Dublin · ★4.88 · 1 bedroom · 1.5 shared baths',
 'Home in Lucan · ★4.93 · 1 bedroom · 1 bed · 1.5 shared baths']

### 13) Which ameneties are available in the first AirBnB listing?

In [14]:
data[0]['amenities']

['Luggage dropoff allowed',
 'Long term stays allowed',
 'Washer',
 'Oven',
 'Stove',
 'Coffee maker',
 'Bed linens',
 'Backyard',
 'Refrigerator',
 'Room-darkening shades']

### 14) Edit the first listing. Add 'wifi' to the amenities.

In [15]:
data[0]['amenities'].append('wifi')
data[0]['amenities']

['Luggage dropoff allowed',
 'Long term stays allowed',
 'Washer',
 'Oven',
 'Stove',
 'Coffee maker',
 'Bed linens',
 'Backyard',
 'Refrigerator',
 'Room-darkening shades',
 'wifi']

### Supplementary Challeneges

If you have finished and know how to write user-defined functions using the def keyword, have a go at these. If you have finished and don't know how to write functions, let your instructor know.

### 15) Write a function that will take in one whole listing and print its price. Name the function get_price. 
Call your function with the argument data[0] to check it works.

Call it a second time with the argument data[1].


In [16]:
def get_price(listing):
    print(listing['price'])

get_price(data[0])
get_price(data[1])


$149.00
$74.00


### 16) Use a for loop to call the function get_price on the first 10 listings in the data. This should print the price of each listing.

In [17]:
for i in range(10):
    get_price(data[i])

$149.00
$74.00
$73.00
$50.00
$82.00
$451.00
$436.00
$80.00
$65.00
$30.00


### 17) Create functionality that will recieve the name of an amenity and print out all the listings that have that amenity.

 Test your function by calling it with the argument 'Pool table'
 
 Test again your function by calling it with the argument 'Gym'


In [18]:
def get_listings_with_amenity(amenity):
    print(f"Listings with {amenity}:")
    for listing in data:
        if amenity in listing['amenities']:
            print(listing['name'])
    

get_listings_with_amenity("Gym")


Listings with Gym:
Rental unit in Dublin · 1 bedroom · 1 bed · 1 bath
Place to stay in Dublin · ★4.63 · 1 bedroom · 1 bed · 1 bath
Rental unit in Dublin · 2 bedrooms · 3 beds · 1.5 baths
Rental unit in Dublin · ★5.0 · 2 bedrooms · 2 beds · 2 baths
Home in Ballsbridge · ★5.0 · 2 bedrooms · 2 beds · 2 baths
Rental unit in Dublin 8 · ★5.0 · 1 bedroom · 1 bed · 1 shared bath
Home in Dublin · ★4.45 · 1 bedroom · 1 bed · 1 shared bath
Townhouse in Dublin 8 · 1 bedroom · 1 bed · 1 shared bath
Rental unit in Dublin 8 · ★3.92 · 4 bedrooms · 4 beds · 4 baths
Rental unit in Dublin · 1 bedroom · 1 bed · 1.5 baths
Townhouse in Swords · ★4.87 · 1 bedroom · 2 beds · 1 shared bath
Townhouse in Swords · ★5.0 · 1 bedroom · 1 bed · 1 shared bath
